Tasks/AzureRmWebAppDeployment/jsonvariablesubs.ts

```diff
@@ -0,0 +1,84 @@
+import tl = require('vsts-task-lib/task');
+import path = require('path');
+import fs = require('fs');
+
+function createEnvTree() {
+    var envVarTree = {
+        value: null,
+        isEnd: false,
+        child: {}
+    };
+    var envVariables = tl.getVariables();
+    if(envVariables.length == 0) {
+        throw new Error(tl.loc('UpgradeAgent'));
+    }
+    for(let envVariable of envVariables) {
+        var envVarTreeIterator = envVarTree;
+        var envVariableName = (envVariable.name).split('.');
+        
+        for(let variableName of envVariableName) {
+            if(envVarTreeIterator.child[variableName] === undefined) {
+                envVarTreeIterator.child[variableName] = {
+                    value: null,
+                    isEnd: false,
+                    child: {}
+                };
+            }
+            envVarTreeIterator = envVarTreeIterator.child[variableName];
+        }
+        envVarTreeIterator.isEnd = true;
+        envVarTreeIterator.value = envVariable.value;
+    }
+    return envVarTree;
+}
+
+function checkEnvTreePath(jsonObjectKey, index, jsonObjectKeyLength, envVarTree) {
+    if(index == jsonObjectKeyLength) {
+        return envVarTree;
+    }
+    if(envVarTree.child[ jsonObjectKey[index] ] === undefined) {
+        return undefined;
+   }
+    return checkEnvTreePath(jsonObjectKey, index + 1, jsonObjectKeyLength, envVarTree.child[ jsonObjectKey[index] ]);
+}
+
+function jsonVariableSubstitutionUtil(jsonObject, envObject) {
+    for(var jsonChild in jsonObject) {
+        var jsonChildArray = jsonChild.split('.');
+        var resultNode = checkEnvTreePath(jsonChildArray, 0, jsonChildArray.length, envObject);
+        if(resultNode != undefined) {
+            if(resultNode.isEnd && typeof jsonObject[jsonChild] !== "object") {
+                jsonObject[jsonChild] = resultNode.value;
+            }
+            else {
+                jsonVariableSubstitutionUtil(jsonObject[jsonChild], resultNode);
```

Comment:
You can prevent one recursive call by adding small condition in case envTree is ended but jsonObject has child